### ISSUE TITLE
Level4 4-2

### ISSUE THREAD

**Issue by brandonrhan** [nus-cs2030/2122-s1#229]
<img width="1082" alt="截屏2021-11-11 下午3 23 25" src="https://user-images.githubusercontent.com/92209052/141255969-0c8c5a52-8994-400c-b7b0-9b6cd801300d.png">

Hi i dont understand the test-case 4-2. As given in the sample output, customer 8 is the no.11 done so server2 is supposed to rest for 25.46s, but why server2 is serving customer17 at 15.524?

**Comment by mewrion** [comment]
Self checkout shouldn't have any rest times I believe! You can see Russel's visualization [#115](https://github.com/nus-cs2030/2122-s1/issues/115)   on how the serve time is computed

**Comment by brandonrhan** [comment]
> Self checkout shouldn't have any rest times I believe! You can see @RussellDash332 visualization [#115](https://github.com/nus-cs2030/2122-s1/issues/115) on how the serve time is computed

wow thanks a lot, I get it now

**Comment by ff1Zzd** [comment]
Just to add on, you could try to skip the restEvent or skip taking rest time (depends on how you implement it) if you detect that the server is a self check out.

**Comment by albertsutz** [comment]
if you use a queue for the restTimes, then what you can do is to poll only when it is a human server :)

If it is a self checkout, don't poll the rest times

**Comment by kxshxsh** [comment]
So I drafted my longest order to be:
Arrive -> Wait -> Serve -> Rest -> Done

So in simulator, I have reference to current event and the next event. I would see if the next event to be executed is Rest. Then from the current event I would see if type of server is human or self. If it's a self check out then I would pass Rest with Rest period- 0.000 otherwise, I would pass Rest with the spontaneously allocated rest time.

**Comment by PokezardVGC** [comment]
To add to this, can I assume that if I take in the rest timings as an array, we only progress to the next index of the array if the current done event is a human server? 
Had an alternative interpretation where we still continue onto the next index of the array, just that if it is a self-checkout, we do not use the rest timings but still traverse the array

**Comment by albertsutz** [comment]
@PokezardVGC 
yes you can (for the first question)
if it is a self-checkout server, then don't traverse the array. This will be useful for Level 5 also

In which, you only generate the next restTime IF the DONE event is from a human server

**Comment by kxshxsh** [comment]
> To add to this, can I assume that if I take in the rest timings as an array, we only progress to the next index of the array if the current done event is a human server? Had an alternative interpretation where we still continue onto the next index of the array, just that if it is a self-checkout, we do not use the rest timings but still traverse the array

Yes to your first question- only progress when it is a human server. I'm not sure if your alternate interpretation would work, let me try to explain why so.
Lets assume they've given N rest times as 1.000 N times.. these N rest times are essentially the rest times 'for human servers' that will be allocated to the the human servers who completed the first N services. Oh lord confusing statement haha I apologise! Just because they gave us N rest times and N customers doesn't mean that after serving every customer, a rest time should be allocated- i think that is where the interpretation can get tough. Its only for the first N human-server-facillitated services:) The first completed N services can also NOT be pre decided so you just allocate the rest times as they come and pass on rest if it is a self check out server:)

Hope this helps! <3

**Comment by PokezardVGC** [comment]
Thanks peeps for the clarifications

**Comment by PokezardVGC** [comment]
Also to clarify, the Self-Checkout counter's queue length is the same as a normal human server's queue length? So let's say we have 2 Self-Checkout counters, the queue length will not be doubled right?

**Comment by optionalemon** [comment]
> Also to clarify, the Self-Checkout counter's queue length is the same as a normal human server's queue length? So let's say we have 2 Self-Checkout counters, the queue length will not be doubled right?

Yep, the max queue length is same as the human server, but note that the queue can now go to two counters to be served!

**Comment by kxshxsh** [comment]
> Also to clarify, the Self-Checkout counter's queue length is the same as a normal human server's queue length? So let's say we have 2 Self-Checkout counters, the queue length will not be doubled right?

Yes! k self check out counters will share only 1 queue which will have the same maximum queue length as all other human servers
